internal/core/data/event_test.go

```diff
@@ -16,32 +16,185 @@ package data
 
 import (
 	"encoding/json"
+	"fmt"
 	"net/http"
 	"net/http/httptest"
 	"os"
 	"strconv"
 	"testing"
 
+	"github.com/edgexfoundry/edgex-go/internal/core/data/messaging"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/db"
 	"github.com/edgexfoundry/edgex-go/internal/pkg/db/memory"
 	"github.com/edgexfoundry/edgex-go/pkg/clients/logging"
+	"github.com/edgexfoundry/edgex-go/pkg/clients/metadata/mocks"
 	"github.com/edgexfoundry/edgex-go/pkg/models"
 	"github.com/gorilla/mux"
+	"github.com/stretchr/testify/mock"
+	"gopkg.in/mgo.v2/bson"
+	"sync"
+	"time"
 )
 
 var testEvent models.Event
 var testRoutes *mux.Router
 
+const (
+	TEST_DEVICE_NAME string = "Test Device"
+	TEST_ORIGIN      int64  = 123456789
+)
+
 func TestMain(m *testing.M) {
-	testEvent.Device = "test device"
-	testEvent.Origin = 123456789
-	dbClient = &memory.MemDB{}
-	testEvent.ID, _ = dbClient.AddEvent(&testEvent)
 	testRoutes = LoadRestRoutes()
 	LoggingClient = logger.NewMockClient()
+	mdc = newMockDeviceClient()
+	ep, _ = messaging.NewEventPublisher(messaging.MOCK, messaging.PubSubConfiguration{})
+	chEvents = make(chan interface{}, 10)
 	os.Exit(m.Run())
 }
 
 //Test methods
+func TestCount(t *testing.T) {
+	prepareDB()
+	c, err := count()
+	if err != nil {
+		t.Errorf(err.Error())
+		return
+	}
+
+	if c != 1 {
+		t.Errorf("expected event count 1, received: %s", strconv.Itoa(c))
+		return
+	}
+}
+
+func TestCountByDevice(t *testing.T) {
+	prepareDB()
+	count, err := countByDevice(testEvent.Device)
+	if err != nil {
+		t.Errorf(err.Error())
+		return
+	}
+
+	if count == 0 {
+		t.Errorf("no events found")
+		return
+	}
+}
+
+func TestDeleteByAge(t *testing.T) {
+	prepareDB()
+	time.Sleep(time.Millisecond * time.Duration(5))
+	age := db.MakeTimestamp()
+	count, err := deleteByAge(age)
+	if err != nil {
+		t.Errorf(err.Error())
+		return
+	}
+
+	if count == 0 {
+		t.Errorf("no events found")
+		return
+	}
+}
+
+func TestGetEvents(t *testing.T) {
```

Comment:
I would strongly advise table-based test approach, as described here: https://medium.com/@sebdah/go-best-practices-testing-3448165a0e18 or here: https://www.youtube.com/watch?v=hVFEV-ieeew for testing different cases (and asserting different outcomes).

Also, for assertion I advise [testify](https://github.com/stretchr/testify).

Reply:
Agreed, table-driven testing is something we should be more familiar with. Testify is used in the current code for mocking out certain methods and behaviors, but I'm sure it has more capabilities. If you'd care to provide a sample with some guidance, that'd be great.

Reply:
You can find some examples in edgex-go, like:
https://github.com/edgexfoundry/edgex-go/blob/master/internal/support/logging/criteria_test.go

Reply:
I created an example table driven test for getEvent: https://gist.github.com/feclare/b901d17c7d9498270e10d7baab2d40ef